Cubical/Algebra/CommRing/Integers.agda

```diff
@@ -2,22 +2,132 @@
+      ⟦ sucInt (pos' n +' y) ⟧  ≡⟨ sym $ suc-⟦⟧ _ ⟩
+      sucᵇ  ⟦ pos' n +' y ⟧     ≡[ i ]⟨ sucᵇ $ isHom+ (pos' n) y i ⟩
+      sucᵇ (⟦ pos' n ⟧ + ⟦ y ⟧) ≡⟨ refl ⟩
+      sucᵇ  ⟦ pos' n ⟧ + ⟦ y ⟧  ∎
+    isHom+ (negsuc' zero) y = pred-suc-inj _ _ (λ i → predl (γ i)) where
+      γ : sucᵇ (⟦ negsuc' zero +' y ⟧) ≡ sucᵇ (pred zero + ⟦ y ⟧)
+      γ = suc-⟦⟧ (negsuc' zero +' y) ∙ (λ i → ⟦ (Int.sucInt+ (negsuc' zero) y ∙ Int.+-comm 0 y) i ⟧) ∙ sym (suc-pred ⟦ y ⟧)
+    isHom+ (negsuc' (suc n)) y = (λ i → ⟦ Int.predInt+ (negsuc' n) y (~ i) ⟧) ∙ sym (pred-⟦⟧ (negsuc' n +' y)) ∙ (λ i → pred $ isHom+ (negsuc' n) y i)
+
+    isHom· : ∀ x y → ⟦ x ·' y ⟧ ≡ ⟦ x ⟧ · ⟦ y ⟧
+    isHom· (pos' zero) y i = ⟦ Int.signed-zero (Int.sgn y) i ⟧
+    isHom· (pos' (suc n)) y =
+      ⟦ pos' (suc n) ·' y ⟧      ≡⟨ cong ⟦_⟧ $ Int.·-pos-suc n y ⟩
+      ⟦ y +' pos' n ·' y ⟧       ≡⟨ isHom+ y _ ⟩
+      ⟦ y ⟧ + ⟦ pos' n ·' y ⟧    ≡[ i ]⟨ ⟦ y ⟧ + isHom· (pos' n) y i ⟩
+      ⟦ y ⟧ + ⟦ pos' n ⟧ · ⟦ y ⟧ ≡⟨ (λ i → ⟦ y ⟧ + ·-comm ⟦ pos' n ⟧ ⟦ y ⟧ i) ∙ sym (·-suc ⟦ y ⟧ ⟦ pos' n ⟧) ∙ ·-comm ⟦ y ⟧ _ ⟩
+      sucᵇ ⟦ pos' n ⟧ · ⟦ y ⟧ ∎
+    isHom· (negsuc' zero) y =
+      ⟦ -1 ·'  y ⟧ ≡⟨ cong ⟦_⟧ (Int.·-neg1 y) ⟩
+      ⟦ -'     y ⟧ ≡⟨ sym (neg-⟦⟧ y) ⟩
+        -    ⟦ y ⟧ ≡⟨ sym (·-neg1 ⟦ y ⟧) ⟩
+        -1 · ⟦ y ⟧ ∎
+    isHom· (negsuc' (suc n)) y =
+      ⟦ negsuc' (suc n) ·' y ⟧         ≡⟨ cong ⟦_⟧ $ Int.·-negsuc-suc n y ⟩
+      ⟦ -' y   +'  negsuc' n   ·'  y ⟧ ≡⟨ isHom+ (-' y) _ ⟩
+      ⟦ -' y ⟧ + ⟦ negsuc' n   ·'  y ⟧ ≡[ i ]⟨ ⟦ -' y ⟧ + isHom· (negsuc' n) y i ⟩
+      ⟦ -' y ⟧ + ⟦ negsuc' n ⟧ · ⟦ y ⟧ ≡⟨ cong₂ _+_ (sym (neg-⟦⟧ y)) refl ⟩
+      -  ⟦ y ⟧ + ⟦ negsuc' n ⟧ · ⟦ y ⟧ ≡⟨ (λ i → - ⟦ y ⟧ + ·-comm ⟦ negsuc' n ⟧ ⟦ y ⟧ i) ∙ sym (·-pred ⟦ y ⟧ ⟦ negsuc' n ⟧) ∙ ·-comm ⟦ y ⟧ _ ⟩
+      pred ⟦ negsuc' n ⟧ · ⟦ y ⟧       ∎
+
+    ⟦⟧-isEquiv : isEquiv ⟦_⟧
+    ⟦⟧-isEquiv = isoToIsEquiv (iso ⟦_⟧ bwd fwd-bwd bwd-fwd)
+
+  IntΣraw : TypeWithStr ℓ-zero RawRingStructure
+  IntΣraw = Int , _+'_ , 1 , _·'_
+
+  BiInvInt≃Int-Σraw : BiInvIntΣraw ≃[ RawRingEquivStr ] IntΣraw
+  BiInvInt≃Int-Σraw = ≃[]-swap rawRingUnivalentStr IntΣraw BiInvIntΣraw ((_ , ⟦⟧-isEquiv) , (isHom+ , pres1 , isHom·))
+
+  Int-CommRingAxioms : CommRingAxioms Int (str IntΣraw)
+  Int-CommRingAxioms = transferAxioms {S = RawRingStructure} rawRingUnivalentStr {axioms = CommRingAxioms} (CommRing→CommRingΣ BiInvIntAsCommRing) IntΣraw BiInvInt≃Int-Σraw
+
+  IntΣ : TypeWithStr _ CommRingStructure
+  IntΣ = (typ IntΣraw) , (str IntΣraw) , Int-CommRingAxioms
+
+  IntAsCommRing : CommRing
+  IntAsCommRing = CommRingΣ→CommRing IntΣ
+
+  Int≃BiInvInt-CommRingEquiv : Σ[ e ∈ ⟨ IntAsCommRing ⟩ ≃ ⟨ BiInvIntAsCommRing ⟩ ] CommRingEquiv IntAsCommRing BiInvIntAsCommRing e
+  Int≃BiInvInt-CommRingEquiv .fst = _ , ⟦⟧-isEquiv
+  Int≃BiInvInt-CommRingEquiv .snd = ringequiv pres1 isHom+ isHom·
+
+  Int≡BiInvInt-AsCommRing : IntAsCommRing ≡ BiInvIntAsCommRing
+  Int≡BiInvInt-AsCommRing = CommRingPath _ _ .fst Int≃BiInvInt-CommRingEquiv
```

Comment:
I now see what is going on here. The problem is that `transferAxioms` doesn't work very well with records... @ecavallo Any thoughts on how to improve this?

Ideally all of the Sigma type definitions and conversions should disappear so that the user can only work with the record definitions.

Reply:
Didn't see this for a long time :sweat_smile:

It could be a good idea to have records for raw structures that are then included in the main structure. Take monoids for example:
```
record RawMonoidStr (A : Type ℓ) : Type (ℓ-suc ℓ) where
  field
    ε        : A
    _·_      : A → A → A

  infixl 7 _·_

record IsMonoid {A : Type ℓ} (raw : RawMonoidStr A) : Type ℓ where

  open RawMonoidStr raw

  field
    isSemigroup : IsSemigroup _·_
    identity    : (x : A) → (x · ε ≡ x) × (ε · x ≡ x)

  open IsSemigroup isSemigroup public

record MonoidStr (A : Type ℓ) : Type (ℓ-suc ℓ) where

  field
    raw : RawMonoidStr A
    isMonoid : IsMonoid raw

  open RawMonoidStr raw public
  open IsMonoid isMonoid public

Monoid : Type (ℓ-suc ℓ)
Monoid = TypeWithStr _ MonoidStr
```
Then at least we can say what the type of the raw structure input to `transferAxioms` should be for an individual structure while staying in the world of records, and it should be possible to write a general `transferAxioms` for records of this kind using reflection. On the other hand it means an additional layer of nesting that might be a pain to work with.

I feel confident one could devise a reflection macro that would make this more convenient even without the change. But the more work takes place on the reflection side, the worse the experience of interactive development tends to be. (At least that's what I've found, but maybe I'm just not writing good macros.)